cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java

```diff
@@ -57,14 +60,26 @@
   private volatile Watch<T> watch;
   private volatile boolean stopped;
 
-  protected AbstractWatcherThread(String threadName, String namespace) {
+  protected AbstractWatcherThread(String threadName, String namespace, KubeResourceType resource) {
     super(threadName);
     this.namespace = namespace;
     this.random = new Random();
 
-    // Resolve <T> to form the concrete type for Watch.Response<T>
-    this.resourceType = TypeToken.of(getClass()).resolveType(
-      AbstractWatcherThread.class.getTypeParameters()[0]).getType();
+    switch (resource) {
```

Comment:
Why need to expand and restrict the type instead of the generic one? The parameterized type `<T>` can be any K8s entity.

Reply:
Previously this class was extended by DeploymentWatcher as
```
class DeploymentWatcher extends AbstractWatcherThread<V1Deployment>
```
This allowed the parameter type to resolve to V1Deployment correctly. However since implementation of StatefulSetWatcher will be same I changed the DeploymentWatcher as

```
class KubeResourceWatcher<T> extends AbstractWatcherThread<T>
```

This causes parameter type to resolve to `T` and hence fails which is why I had to change it to switch case. To keep it generic, we can have `DeploymentWatcher` and `StatefulSetWatcher` however most of code will be duplicate.

Any other option?